### PR TITLE
[WIP] Use truncated gaussian distribution for entropy calculation

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Trying to use clipped gaussian entropy instead of gaussian entropy.
The advantage is that the maximum entropy no longer corresponds to stdev = inf, with this method, the maximum entropy is reached for mean = 0, stdev = 1.6433039903640747

The most spread out distribution of action will look like this : 
Screen Shot 2021-02-17 at 12 37 34

instead of this : 
Screen Shot 2021-02-17 at 12 38 29

Note : In this PR, I multiply the normalization constant Z by 0.95. If I do not, the max entropy distribution corresponds to stdev ~ 80

This code can help see what values of mean and stdev corresponds to max entropy : 

```
import abc
from typing import List
from mlagents.torch_utils import torch, nn
import numpy as np
import math
from mlagents.trainers.torch.layers import linear_layer, Initialization
from mlagents.trainers.torch.distributions import GaussianDistInstance



for m in [0]:
    for s in [0.01,0.1,0.5,1,2,6,12,20, 50, 70, 80,100, 1000]:
        mean = torch.Tensor([[m]])
        std = torch.Tensor([[s]])
        a = GaussianDistInstance(mean, std)
        print(m, s, a.entropy().item())

m = nn.Parameter(torch.zeros(1, 1, requires_grad=True))
s = nn.Parameter(torch.ones(1, 1, requires_grad=True))

optimizer = torch.optim.Adam(
            [m, s], lr=1e-3
        )

for i in range(100000):
    a = GaussianDistInstance(m, s)
    loss = - torch.mean(a.entropy())

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if i % 100 == 0:
        print(i, m.item(), s.item(), a.entropy().item())
```

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

#4866 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/distributions.py [hunk 1]
```diff
@@ -73,11 +73,50 @@ def pdf(self, value):
...
+        # The entropy is calculated using the clipped gaussian entropy:
+        # https://en.wikipedia.org/wiki/Truncated_normal_distribution
+        min_bound = -1
+        max_bound = 1
+        alpha = (min_bound - self.mean) / (self.std + EPSILON)
+        beta = (max_bound - self.mean) / (self.std + EPSILON)
+
+        # 0.95 is a computation trick to keep the maximum sigma within bounds
+        Z = (
+            0.95
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/distributions.py:85]: This multiplication by 0.95 is arbitrary. need to investigate more

**Comment by vincentpierre** [comment]
It seems using the truncated gaussian distribution is not good enough. We need a clipped Gaussian distribution entropy, but I cannot find an easy way to implement its calculation. Same goes with TanhGaussian, the entropy calculation has not been done and probably does not have a non-integral solution.

**Comment by vincentpierre** [comment]
Here is a plot of the Entropy for the Gaussian and for the TanhGaussian 

Entropy as funtion of mu and sigma

The two are different, the Gaussian distribution's entropy is maximum when the standard deviation is infinite and is independent of the mean. On the other hand, the entropy of the TanhGaussian is maximum for a mean 0 and a finite standard deviation (which is to be expected)

The exact equation or even an approximating the entropy of the TanhGaussian is rather hard to compute and there does not seem to be a lot of benefits in being exact in the entropy calculation (a lot of RL algorithms use clipped gaussian which also has a different entropy than a regular gaussian). Therefore I will close this now and add this issue on our backlog.
